File: apps/demo-storefront/frontend/src/lib/proof.ts
/**
 * Honest wording for recommendation provenance.
 *
 * Only `proofStatus` (computed by the storefront server from configuration and the
 * upstream response) drives user-facing claims. "For Maya" / "Model-backed" appear
 * only when the status is `model_verified`; the raw strategy string is shown solely
 * in the technical disclosure, labelled as raw.
 */
import type { ProofStatus, Provenance } from "./types";

export type ProofTone = "verified" | "preview" | "fallback";

export function proofTone(status: ProofStatus): ProofTone {
  if (status === "model_verified") return "verified";
  if (status === "catalog_fallback") return "fallback";
  return "preview";
}

export function proofColor(status: ProofStatus): string {
  return { verified: "var(--verified)", preview: "var(--preview)", fallback: "var(--ink-2)" }[proofTone(status)];
}

export function proofIcon(status: ProofStatus): string {
  return { verified: "✓", preview: "!", fallback: "◇" }[proofTone(status)];
}

export function proofPillLabel(status: ProofStatus): string {
  return { verified: "Verified personalization", preview: "Serving preview · unverified", fallback: "Catalog fallback" }[proofTone(status)];
}

/** Heading over a shelf. Never claims personalisation unless verified. */
export function shelfHeading(status: ProofStatus | null, personaName: string, variant: "home" | "related" | "continue" = "home"): string {
  if (status === "model_verified") return variant === "home" ? `For ${personaName}` : variant === "related" ? "You may also like" : "Continue exploring";
  if (status === "catalog_fallback") return variant === "home" ? "From the catalog" : variant === "related" ? "More from the catalog" : "Continue exploring";
  return variant === "home" ? "Recommended" : variant === "related" ? "You may also like" : "Continue exploring";
}

export function provenanceLine(prov: Provenance | null, personaName: string, isGuest: boolean): string {
  if (!prov) return "Waiting for GraphRec…";
  const version = prov.modelVersionId ? ` · version ${prov.modelVersionId.slice(0, 8)}` : "";
  switch (prov.proofStatus) {
    case "model_verified":
      return `Model-backed for ${personaName}${version}`;
    case "catalog_fallback":
      return isGuest ? "Catalog fallback · new visitor, no history" : `Catalog fallback · showing as ${personaName}`;
    default:
      return `GraphRec serving preview${version} · showing as ${personaName}`;
  }
}

export function proofExplainer(status: ProofStatus): string {
  switch (status) {
    case "model_verified":
      return "Rankings are user-specific, repeat exactly, and this model version passed the verification script.";
    case "catalog_fallback":
      return "GraphRec answered from the catalog (newest active products), not from a model. Labelled as a fallback.";
    default:
      return "The serving path is live and returns a model version, but the query is not yet user-specific. Not presented as personalization.";
  }
}

/** Rows for the "Why am I seeing this?" disclosure. */
export function whyRows(prov: Provenance): { label: string; value: string }[] {
  return [
    { label: "Request ID", value: prov.requestId },
    { label: "Raw strategy (backend)", value: prov.rawStrategy },
    { label: "Fallback", value: String(prov.fallbackUsed) + (prov.fallbackUsed ? ` (${prov.fallbackTier})` : "") },
    { label: "Model version", value: prov.modelVersionId ?? "n/a" },
    { label: "Proof status", value: prov.proofStatus },
    { label: "Excluded products", value: String(prov.excludedCount) },
    { label: "Returned but not shown", value: String(prov.omittedProductCount) },
    { label: "Impression event", value: prov.impressionEventId ?? "—" },
  ];
}

File: apps/demo-storefront/app/proof.py
"""Honest provenance: compute ``proofStatus`` and cross-persona similarity.

The raw ``strategy`` GraphRec returns never drives a user-facing claim. A result is
``model_verified`` only when the operator has recorded a passed verification for
the *exact* model version that produced it.
"""

from __future__ import annotations

from itertools import combinations
from typing import Dict, List, Optional, Sequence, Tuple

from .schemas import PairSimilarity, ProofStatus, Provenance


def proof_status(
    *,
    fallback_used: bool,
    model_version_id: Optional[str],
    verified: bool,
    verified_version_id: str,
) -> ProofStatus:
    if fallback_used:
        return "catalog_fallback"
    if verified and model_version_id and verified_version_id and model_version_id == verified_version_id:
        return "model_verified"
    return "serving_preview"


def provenance(
    recs: object,
    *,
    verified: bool,
    verified_version_id: str,
    excluded_count: int,
    omitted_product_count: int,
    impression_event_id: Optional[str] = None,
) -> Provenance:
    """Build the provenance block from an SDK ``Recommendations`` result."""

    model_version_id = getattr(recs, "model_version_id", None)
    model_version_id = str(model_version_id) if model_version_id else None
    fallback_used = bool(getattr(recs, "fallback_used", False))
    return Provenance(
        request_id=str(getattr(recs, "request_id")),
        model_version_id=model_version_id,
        raw_strategy=str(getattr(recs, "strategy", "")),
        fallback_used=fallback_used,
        fallback_tier=str(getattr(recs, "fallback_tier", "")),
        proof_status=proof_status(
            fallback_used=fallback_used,
            model_version_id=model_version_id,
            verified=verified,
            verified_version_id=verified_version_id,
        ),
        excluded_count=excluded_count,
        omitted_product_count=omitted_product_count,
        impression_event_id=impression_event_id,
    )


def pairwise_similarity(rankings: Dict[str, Sequence[str]], names: Dict[str, str]) -> List[PairSimilarity]:
    pairs: List[PairSimilarity] = []
    for a, b in combinations(rankings.keys(), 2):
        sa, sb = set(rankings[a]), set(rankings[b])
        union = len(sa | sb)
        pairs.append(
            PairSimilarity(a=names[a], b=names[b], overlap=len(sa & sb), jaccard=round(len(sa & sb) / union, 2) if union else 0.0)
        )
    return pairs


def distinct_orderings(rankings: Dict[str, Sequence[str]]) -> int:
    return len({tuple(r) for r in rankings.values()})


def shared_across_all(rankings: Dict[str, Sequence[str]]) -> set:
    sets = [set(r) for r in rankings.values() if r]
    return set.intersection(*sets) if sets else set()


def gate(
    *,
    known_rankings: Dict[str, Sequence[str]],
    any_model_version: bool,
    all_repeatable: Optional[bool],
    any_fallback_for_known: bool,
    verified_status_everywhere: bool,
) -> Tuple[bool, str]:
    """The model-proof gate, mirroring ``scripts/verify_personalization.py``.

    Returns ``(passed, reason)``. The reason is written for a presenter to read aloud.
    """

    if not known_rankings:
        return False, "No shopper with history was compared."
    if not any_model_version:
        return False, "No model version was returned, so there is nothing to verify."
    if any_fallback_for_known:
        return False, "At least one shopper with history received the catalog fallback."
    if all_repeatable is False:
        return False, "Repeating the same request changed the ordering, so results are not stable."
    if distinct_orderings(known_rankings) < len(known_rankings):
        return False, (
            "Shoppers with different histories received the same ranking. The serving path is live "
            "but the query is not yet user-specific."
        )
    if not verified_status_everywhere:
        return False, (
            "Rankings differ per shopper, but this model version has not been recorded as verified "
            "(MODEL_PROOF_VERIFIED / MODEL_PROOF_VERSION_ID). Run scripts/verify_personalization.py."
        )
    return True, "Rankings differ per shopper, repeat exactly, and carry a verified model version."
